zap/src/main/java/org/parosproxy/paros/core/scanner/HostProcess.java

```diff
@@ -317,7 +319,7 @@ public void run() {
                         startNode,
                         true,
                         node -> {
-                            if (canScanNode(node)) {
+                            if (canScanNode(node) && bestNodeForUrl(node)) {
```

Comment:
This check will not work properly for all cases, not all nodes are included by default (e.g. out of scope, filters, not recurse) so we can't discard the temporary node beforehand. Taking the example in the issue, no nodes are scanned if you select the node "3" and disable recurse.

It might be safer to de-duplicate once we have all of them.

Reply:
I forgot that case. I believe it can be solved by adding the check canScanNode to the sibling
```diff
...
            if (!sibling.isSameAs(node)
+                   && canScanNode(sibling)
                    && sibling.getMethod().equals(nodeMethod)
...
``
What do you think?

Reply:
That would address most part of it (e.g. out of scope, filters) but still not the above recurse case, the `canScanNode` does not tell if a node is included in the nodes to be scanned, that's done while traversing (to not check more nodes than really needed).

Reply:
True, I will do it after collecting all history nodes as you suggested.

Reply:
It can also be de-duplicated as they are added, use a set and resolve the duplication immediately (might be simpler).

Reply:
Yes, it's simpler :), I used a dict to keep track of the locations where the requests were stored.